dbt/include/athena/macros/materializations/snapshots/snapshot.sql

```diff
@@ -12,365 +11,112 @@
-    {% if table_type=='iceberg' %}
-        {% set temp_relation = make_temp_relation(target_relation) %}
-
-        {% set select = snapshot_staging_table(strategy, sql, target_relation, table_type) %}
-
-        {% call statement('build_snapshot_staging_relation') %}
-            {{ create_table_as(True, temp_relation, select) }}
-        {% endcall %}
-
-        {% do return(temp_relation) %}
-    {%else%}
-        {%- set tmp_identifier = target_relation.identifier ~ '__dbt_tmp' -%}
-
-        {%- set tmp_relation = api.Relation.create(identifier=tmp_identifier,
-                                                      schema=target_relation.schema,
-                                                      database=target_relation.database,
-                                                      type='table') -%}
-
-        {% do adapter.drop_relation(tmp_relation) %}
-
-        {%- set select = snapshot_staging_table(strategy, sql, target_relation, table_type) -%}
-
-        {% call statement('build_snapshot_staging_relation') %}
-            {{ create_table_as(False, tmp_relation, select) }}
-        {% endcall %}
-
-        {% do return(tmp_relation) %}
-    {% endif %}
+{% macro iceberg_snapshot_merge_sql(target, source, insert_cols) %}
+    {%- set insert_cols_csv = insert_cols | join(', ') -%}
+    {%- set src_columns = [] -%}
+    {%- for col in insert_cols -%}
+      {%- do src_columns.append('dbt_internal_source.' + col) -%}
+    {%- endfor -%}
+    {%- set src_cols_csv = src_columns | join(', ') -%}
+
+    merge into {{ target }} as dbt_internal_dest
+    using {{ source }} as dbt_internal_source
+    on dbt_internal_source.dbt_scd_id = dbt_internal_dest.dbt_scd_id
+
+    when matched
+     and dbt_internal_dest.dbt_valid_to is null
+     and dbt_internal_source.dbt_change_type in ('update', 'delete')
+        then update
+        set dbt_valid_to = dbt_internal_source.dbt_valid_to
+
+    when not matched
+     and dbt_internal_source.dbt_change_type = 'insert'
+        then insert ({{ insert_cols_csv }})
+        values ({{ src_cols_csv }})
 {% endmacro %}
 
+
 {#
-    Update the dbt_valid_to and is_current_record for
-    snapshot rows being updated and create a new temporary table to hold them
+    Create a new temporary table that will hold the new snapshot results.
+    This table will then be used to overwrite the target snapshot table.
 #}
-
-{% macro athena__create_new_snapshot_table(strategy, strategy_name, target, source) %}
-    {%- set tmp_identifier = target.identifier ~ '__dbt_tmp_1' -%}
-
-    {%- set tmp_relation = adapter.get_relation(database=target.database, schema=target.schema, identifier=tmp_identifier) -%}
-
-    {%- set target_relation = api.Relation.create(identifier=tmp_identifier,
-      schema=target.schema,
-      database=target.database,
-      type='table') -%}
+{% macro hive_create_new_snapshot_table(target, source, insert_cols) %}
```

Comment:
This is actually the major change of this PR. This is some sql logic to replicate the `MERGE INTO` statements of dbt-core. The output should be exactly the same as the MERGE INTO output, with the main difference that the results are first stored in a temporary staging table, and only then moved to the final location.